bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -3402,6 +3403,8 @@ void ReplicaImp::onSeqNumIsStable(SeqNum newStableSeqNum, bool hasStateInformati
     IControlHandler::instance()->onStableCheckpoint();
     ControlStateManager::instance().wedge();
   }
+  auto ret = std::async(std::launch::async,
```

Comment:
The way the callback is called (`std::async`) is implementation dependent, so only `DbCheckpointManager::instance().onStableCheckPoint(newStableSeqNum);` should be called.
Also sorry for being not very clear about a callback registry.
I suggested adding `concord::util::CallbackRegistry<uint64_t> onSeqNumStable_;` along with  `concord::util::CallbackRegistry<uint64_t> onViewNumCallbacks_;` to `ReplicaImp`and not inside the `DbCheckpointManager`
Personally I don't see much value in having `DbCheckpointManager::addOnStableSeqNum()`, since it basically is not a callback, but a known to  `DbCheckpointManager` function. And adding it in `ReplicaImp::ReplicaImp()` doesn't make much sense.

Reply:
Fixed